### PR TITLE
Allow integer log levels

### LINKED ISSUE #2: Stop assuming `LogLevel::*` levels, because `\Psr\Log\LoggerInterface::log()`does not require it either
👋 

We're looking into adopting this in Drupal over at https://www.drupal.org/project/drupal/issues/3321905.

However, we ran into a blocker. We have to do write code like
```
$this->assertTrue($logger->hasRecordThatContains('[name] => AAA 8.x-5.x', (string) RfcLogLevel::ERROR));
```

… because this package assumes the level is always a `string`, specifically one of `LogLevel::*` (although that's not enforced, but just prescribed by the documentation).

But `\Psr\Log\LoggerInterface::log()` prescribes only `mixed   $level`. And the Drupal logging implementation (`\Drupal\Core\Logger\LoggerChannel::log()`) stores _integers_, not _strings_:
```
    if (is_string($level)) {
      // Convert to integer equivalent for consistency with RFC 5424.
      $level = $this->levelTranslation[$level];
    }
```

… which means that we unfortunately cannot actually use `colinodell/psr-testlogger`, except in the awkward way shown above. It also means that we can _never_ use the nicer accessors such as `hasErrorThatContains()` etc.

Therefore I propose to optionally allow specifying a mapping of `LogLevel::*` strings to alternative values (in our case: RFC 5424-compliant integers). This would present no BC break; it would be entirely optional.

### PR DESCRIPTION
**Pull request by jonathanjfshaw**
This is an alternative implementation to https://github.com/wimleers/psr-testlogger/pull/1

Fixes #2

### CODE REVIEW

**Comment by jonathanjfshaw** [comment]
I've used union types for string|int, which require php 8. Not sure whether you want to make a php 8 release, or drop the explicit type.

File: src/TestLogger.php [hunk 1]
```diff
@@ -65,6 +65,27 @@ final class TestLogger extends AbstractLogger
     /** @var array<int|string, array<int, array<string, mixed>>> */
     public array $recordsByLevel = [];
 
+    /** @var array<string, string|int> */
+    public array $levelMap = [
```
Comment (colinodell) [inline src/TestLogger.php]: I think I'd prefer to make this `private` for now so that outside code can't mutate it after instantiation.

```suggestion
    private array $levelMap = [
```

File: src/TestLogger.php [hunk 2]
```diff
@@ -65,6 +65,27 @@ final class TestLogger extends AbstractLogger
...
+        LogLevel::ERROR => LogLevel::ERROR,
+        LogLevel::WARNING => LogLevel::WARNING,
+        LogLevel::NOTICE => LogLevel::NOTICE,
+        LogLevel::INFO => LogLevel::INFO,
+        LogLevel::DEBUG => LogLevel::DEBUG,
+    ];
+
+    /**
+     * @param array<string, string|int>|null $levelMap
+     *   Keys are LogLevel::*, values are alternative strings or integers used as log levels in the SUT.
```
Comment (colinodell) [inline src/TestLogger.php]: We can type-hint the allowed keys to phpstan like this:

```suggestion
     * @param array<LogLevel::*, string|int>|null $levelMap
     *   Keys are LogLevel::*, values are alternative strings or integers used as log levels in the SUT.
```

File: src/TestLogger.php [hunk 3]
```diff
@@ -153,10 +174,11 @@ public function hasRecordThatPasses(callable $predicate, string $level): bool
      */
     public function __call(string $method, array $args): bool
     {
-        if (\preg_match('/(.*)(Debug|Info|Notice|Warning|Error|Critical|Alert|Emergency)(.*)/', $method, $matches) > 0) {
+        $levelNames = \implode('|', \array_map('ucfirst', \array_keys($this->levelMap)));
+        if (\preg_match('/(.*)(' . $levelNames . ')(.*)/', $method, $matches) > 0) {
```
Comment (colinodell) [inline src/TestLogger.php]: I'm guessing the idea behind this change is that a user may have instantiated the `TestLogger` without defining all of those levels in their map?  If so, I think my preference would be to revert this change and add some validation to the constructor to ensure that the map contains all log levels.  That would ensure that all of the magic methods defined on lines 15-54 are always available and guaranteed to never throw a `BadMethodCallException`.

**Review by colinodell — COMMENTED** [review]
Overall I think this approach is on the right track!  Besides a few nits, I think the only other thing this PR needs is some tests showing that the level mapping works properly.

With regards to PHP 7.4: I'm fine with dropping support for that and bumping the minimum version to `^8.0`.  We can then release this as a minor version update since the changes would be backward-compatible.

File: src/TestLogger.php (hunk 3 above)
Comment (jonathanjfshaw) [reply src/TestLogger.php]: Actually this was motivated by DRY. It smelled to me to repeat the name information twice in the class. I have kept this code, but added the requested validation to the __construct() method. Happy to remove this code if you prefer the simplicity of explicit repetition instead of DRY.

**Comment by jonathanjfshaw** [comment]
All tests and checks are green now, I think this is good to go. Only outstanding question is the DRY vs explicit issue in your review point discussed above.

**Comment by jonathanjfshaw** [comment]
@colinodell Any chance of this getting merged? Drupal are still keen to adopt this package.

**Comment by colinodell** [comment]
Sorry for the delay here! Life has been quite busy lately :)

After giving this a thorough review, I think I'd like to remove the `$levelMap` functionality.  On the surface it seems potentially useful, but it leads to unexpected behavior like this:

```php
$levelMap = [
    LogLevel::EMERGENCY => 0,
    LogLevel::ALERT => 1,
    LogLevel::CRITICAL => 2,
    LogLevel::ERROR => 3,
    LogLevel::WARNING => 4,
    LogLevel::NOTICE => 5,
    LogLevel::INFO => 6,
    LogLevel::DEBUG => 7,
];

$logger = new TestLogger($levelMap);

$logger->debug('foo');
$logger->log(LogLevel::DEBUG, 'bar');
$logger->log(7, 'baz');

$this->assertTrue($logger->hasDebugThatContains('foo')); // FAILS
$this->assertTrue($logger->hasDebugThatContains('bar')); // FAILS
$this->assertTrue($logger->hasDebugThatContains('baz')); // passes
```

After re-reading [section 1.1 of PSR-3](https://www.php-fig.org/psr/psr-3/#11-basics), it seems like `7` and `LogLevel::DEBUG` should pass into the logger as completely distinct, unrelated values.  But the level map conflates the two.  I can see how that mapping might be useful in a system that doesn't use `LogLevel` constants at all, but from this library's perspective, I think we should avoid the assumption that people will always use ONLY the constants or ONLY custom levels.  So while I'm sure we could find a smart implementation that would work in all cases, I feel like doing so would contradict the intent of the spec.

Yes, this would mean that Drupal would not be able to leverage the magic methods.  But all of the other methods would work just fine with `int` log levels, which I think would be sufficient.

WDYT?

If you think this is fine, I'd be happy to make the final changes to the PR and get this merged/released ASAP :)

**Comment by jonathanjfshaw** [comment]
Thanks for looking @colinodell . I think removing the level map and not being able to use the magic methods will cause Drupal no problems at all.

I've asked @wimleers the Drupal expert who detected the present issue to give this MR a final check.

**Comment by colinodell** [comment]
Sounds good! I've updated the code in this PR accordingly and look forward to hearing the results of that final check :)

**Review by colinodell — APPROVED** [review]
Ready to merge once feedback is received

File: composer.json [hunk 4]
```diff
@@ -19,7 +19,7 @@
         "source": "https://github.com/colinodell/psr-testlogger"
     },
     "require": {
-        "php": "^7.4 || ^8.0",
+        "php": "^8.0",
```
Comment (wimleers) [inline composer.json:22]: According to `semver` this means it should be version `2.0.0`? 🤔

**Comment by wimleers** [comment]
Yep, that seems reasonable to me too!

This PR looks fine to me, with the sole exception that bumping the required PHP version in anything else than a major version release AFAICT violates https://semver.org/ 😅 If you feel strongly about not tagging a new major version, fine, but … I'm fairly sure this would make the Drupal maintainers feel uneasy because if this happens again at some point in the future, then it technically breaks backwards compatibility for Drupal 🤓

**Review by wimleers — APPROVED** [review]

File: composer.json (hunk 4 above)
Comment (catch56) [reply composer.json:22]: semver doesn't tie you to a particular PHP version support cycle, it's up to the project to determine what counts as a breaking API change and document it, and if PHP version increases don't count, then they don't.

Comment (colinodell) [reply composer.json:22]: Correct, a bump in dependencies is not a BC break - see https://github.com/semver/semver/blob/d460299638d3d2af6fcf809285dc6a469e15742a/semver.md#what-should-i-do-if-i-update-my-own-dependencies-without-changing-the-public-api

The only reason we might need to release this as a new major version would be if we BC in the public API. Although the API signature of `TestLogger` _did_ change, all of those changes are backward-compatible:

 - The class is `final`, so we don't need to worry about child classes suddenly becoming incompatible
 - The only type changes were parameter type widening. The new `string|int` types will continue to accept `string` inputs with no changes needed by end users.

> I'm fairly sure this would make the Drupal maintainers feel uneasy because if this happens again at some point in the future, then it technically breaks backwards compatibility for Drupal

This shouldn't be an issue as Composer will prevent the installation of newer versions with incompatible dependencies.

For example, this change will be released as v1.2.0, so I imagine Drupal will use a constraint like `^1.2`.  If I were to release v1.3.0 in the future with a new minimum PHP version of `8.2.0`, Composer would still allow v1.2.0 on PHP 8.0 and 8.1 - only users on 8.2 could install v1.3.0.  And because it would be a minor version bump and we follow semver, there would be no breaking changes to the API.  This means that both versions would function identically as far as Drupal is concerned.

Bumping dependencies (including minimum PHP versions) in a minor release is done by many other large PHP projects, including Doctrine: https://www.doctrine-project.org/2017/07/25/php-7.1-requirement-and-composer.html

**TL;DR:** Everything will "just work" :)
